hedera-node/hedera-app/src/main/java/com/hedera/node/app/tss/handlers/TssVoteHandler.java

```diff
@@ -53,5 +60,61 @@ public void pureChecks(@NonNull final TransactionBody txn) throws PreCheckExcept
         requireNonNull(context);
+        final var txBody = context.body().tssVoteOrThrow();
+        final var tssBaseStore = context.storeFactory().writableStore(WritableTssStore.class);
+        final TssVoteMapKey tssVoteMapKey = new TssVoteMapKey(
+                txBody.targetRosterHash(), context.creatorInfo().nodeId());
+        if (tssBaseStore.exists(tssVoteMapKey)) {
+            // Duplicate vote
+            return;
+        }
+
+        if (!TssVoteHandler.hasReachedThreshold(txBody, context)) {
+            tssBaseStore.put(tssVoteMapKey, txBody);
+        }
+    }
+
+    /**
+     * Check if the threshold number of votes (totaling at least 1/3 of weight) have already been received for the
+     * candidate roster, all with the same vote byte array.
+     *
+     * @param tssVoteTransaction the TssVoteTransaction to check
+     * @param context the HandleContext
+     * @return true if the threshold has been reached, false otherwise
+     */
+    public static boolean hasReachedThreshold(
+            @NonNull final TssVoteTransactionBody tssVoteTransaction, @NonNull final HandleContext context) {
+        final var rosterStore = context.storeFactory().readableStore(ReadableRosterStore.class);
+
+        final Roster activeRoster = rosterStore.getActiveRoster();
+        if (activeRoster == null) {
+            throw new IllegalArgumentException("No active roster found");
+        }
+        // Get the target roster from the TssVoteTransactionBody
+        final Bytes targetRosterHash = tssVoteTransaction.targetRosterHash();
+
+        // Also get the total active roster weight
+        long activeRosterTotalWeight = 0;
+        // Initialize a counter for the total weight of votes with the same vote byte array
+        long voteWeight = 0L;
+        final var tssBaseStore = context.storeFactory().writableStore(WritableTssStore.class);
+        // For every node in the active roster, check if there is a vote for the target roster hash
+        for (final RosterEntry rosterEntry : activeRoster.rosterEntries()) {
+            activeRosterTotalWeight += rosterEntry.weight();
+            final var tssVoteMapKey = new TssVoteMapKey(targetRosterHash, rosterEntry.nodeId());
+            if (tssBaseStore.exists(tssVoteMapKey)) {
+                final var vote = tssBaseStore.getVote(tssVoteMapKey);
+                // If the vote byte array matches the one in the TssVoteTransaction, add the weight of the vote to the
+                // counter
+                if (vote.tssVote().equals(tssVoteTransaction.tssVote())) {
+                    voteWeight += rosterEntry.weight();
+                }
+            }
+        }
+
+        // Check if the total weight of votes with the same vote byte array is at least 1/3 of the
+        // total weight of the network
+        // Adding a +1 to the threshold to account for rounding errors.
+        return voteWeight >= (activeRosterTotalWeight / 3) + 1;
```

Comment:
This formula still doesn't look right, see `Threshold.java` for the correct strong minority computation

```
part >= (whole / 3) + ((whole % 3 == 0) ? 0 : 1);
```

Reply:
Ah fixed. Thanks @litt3